protos/sigstore_trustroot.proto

```diff
@@ -0,0 +1,78 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+package dev.sigstore.trustroot.v1;
+
+import "sigstore_common.proto";
+
+option go_package = "github.com/sigstore/proto/trustroot/v1";
+option java_package = "dev.sigstore.proto.trustroot.v1";
+option java_multiple_files = true;
+option java_outer_classname = "TrustRootProto";
+
+// TransparencyLogInstance describes the data required from a
+// transparency log, to successfully identify it and perform verification
+// inclusion.
+message TransparencyLogInstance {
+        // The unique identifier for this transparency log.
+        dev.sigstore.common.v1.LogId log_id = 1;
+        // The URI at which the log can be accessed.
+        string uri = 2;
+        dev.sigstore.common.v1.PublicKey public_key = 3;
+        // The time range is half-open and does not include the end
+        // timestamp, i.e [valid_for.start, valid_for.end).
+        dev.sigstore.common.v1.TimeRange valid_for = 4;
+}
+
+// CertificateAuthority enlists the information required to identify which
+// CA to use and perform signature verification.
+message CertificateAuthority {
+        // The root certificate MUST be self-signed, and so the subject and
+        // issuer are the same.
+        dev.sigstore.common.v1.DistinguishedName subject = 1;
+        // The URI at which the CA can be accessed.
+        string uri = 2;
+        // A single chain for a signing certificate.
+        dev.sigstore.common.v1.X509CertificateChain cert_chain = 3;
+        // The time the *entire* chain was valid. This is at max the
+        // longest interval when *all* certificates in the chain where valid,
+        // but it MAY be shorter.
+        // The time range is half-open and does not include the end
+        // timestamp, i.e [valid_for.start, valid_for.end).
+        dev.sigstore.common.v1.TimeRange valid_for = 4;
+}
+
+// TrustedRoot describes the client's complete set of trusted entities.
+// How the TrustedRoot is populated is not specified, but can be a
+// combination of many sources such as TUF repositories, files on disk etc.
+message TrustedRoot {
+        // A set of trusted Rekor servers.
+        repeated TransparencyLogInstance tlogs = 1;
+        // A set of trusted certificate authorites, and any intermediate
+        // certificates they provide.
+        // If a CA is issuing multiple intermediate certificate, each
+        // combination shall be represented as separate chain. I.e, a single
+        // root cert may appear in multiple chains but with different
+        // intermediate and/or leaf certificates.
+        // The certificates are intended to be used for verifying artifact
+        // signatures.
+        repeated CertificateAuthority certificate_authorities = 2;
+        // A set of public keys to trust for artifact signing.
```

Comment:
+1

Reply:
> I think it's sort of on the plate in the long term, but if we don't need it right now, I'd suggest leaving it out

I think this is the right thing to do for now.

> since I'm not sure the correct semantics for this struct yet. e.g. what is the public key designated to sign? I don't think it would be a repeated PublicKey but more a repeated ArtifactPublicKey that contains public keys or identities associated with artifact namespaces or something else.

Such a mapping feels very separate to Sigstore verification to me—it would be very application-specific.

I think I'm agreeing with you, overall, though—a universe of possible public keys to trust feels weird to me; rather, prefer having users pass in a public key *instead of* and Identity at verification time; it's up to them to resolve which key.

Reply:
The way I see it is like this. Lets say you have a build system with more than one steps. Each step (e.g. build, unit test, vulnerability scan etc) relies on a specific key (in typical in-toto fashion). Each step produces a signed attestation. During verification (lets say prior to deployment), you would like to verify that the set of attestations exists based on some policy. For that, the verifier would need to have access to a set of public keys, to be able to verify all the attestations, and that the correct actor signed them. That is the thinking I had in mind. 
Another use case is that the there is a single signing key, but it's rotated on a schedule, and we may want to verify signatures made in the past.

A similar concept is the public keys to Rekor. They are now, managed by a set of `TransparencyLogInstance`s, but the concept is the same.

Reply:
That seems like a legitimate use case, but I still think this is a property of the *individual verification* and not the *Sigstore trust root*. So I'd expect these to go into `ArtifactVerificationOptions`. Does that make sense?

Reply:
> So I'd expect these to go into ArtifactVerificationOptions. Does that make sense?

Yes and no :)

I was under the impression that the input to the artifact verification is always a sub-set of what's availabile in the _Sigstore trust root_?

Reply:
I think of it as analogous to the "identity," which isn't (explicitly) in the trust root but is in the verification options.

Reply:
What are the other use-cases for public keys beyond seeing it as an "identity"? The public keys in the trust root is for verifying artifact signatures, which IMHO falls in this scenario. For other things such as Rekor or the CT, those public keys are managed explicitly within a `TransparencyLogInstance` message, where each key is bound to a specific log.

Reply:
> The public keys in the trust root is for verifying artifact signatures, which IMHO falls in this scenario.

Here's how I've been thinking about it:

- In the keyless setting, user@example.com via accounts.google.com gets a certificate C from Fulcio, then signs artifact X and sticks a record of that signature in the transparency log.
- In the key-full setting, key K signs artifact X and sticks a record of that signature in the transparency log.

To me, "key K" is basically "user@example.com via accounts.google.com + certificate C", with no need for Fulcio. 

From the verifier end, I expect to create policies like:

- all artifacts must be signed by user@example.com
- all artifacts must be signed by key K

So I'd expect to provide "key K" in the same place I'd provide "user@example.com via accounts.google.com", which is the verification options.


-------------------------------------------

There's probably a coherent argument for doing it otherwise, that's just how I've been imagining things.

Reply:
I say we are referring to the same use-case/scenario -- a private key is used as an identity. 

So the way I was thinking the use of the trust root would be like this:
Upon process start the, the trusted key material is gathered (e.g. via TUF), the resulting trust root looks like this:
```JSON
{
...
"public_keys": [{"key_id":"1"}, {"key_id":"2"}, {"key_id":"2"}]
...
}
```
An artifact A is to be signed. The policy is that key with `key_id:2` should be used.
A selection phase happens and a filtered (per artifact) trust root is created based on the global:

```JSON
{
...
"public_keys": [{"key_id":"2"}]
...
}
```

Which is then used by the verification function. 

The same reasoning is what I expect to happen for all the items in the trust root too, but in this case we are talking about public keys. The input to the verify method is a filtered trust root and the options. 

The difference from your example is that the identity when using a certificate is put in the options struct (and the CA in the trust root), whereas the public keys are put in the trust root struct.

Reply:
I guess I understand this a little better: policies often allow users to specify the trusted keys, and this is meant to be an unfiltered collection of global trusted signing keys (with no policy on its usage).

I'm still a little hesitant on including it, but I dont think it needs to be blocked on: I can see it being used: e.g. builder's trust root would include all their builder signing keys. I feel like it should be separate from the other keys here, but I'm not going to block because I'm having trouble arguing that globally trusted material for the ecosystem is distinct from globally trusted material for the artifacts. (Because ecosystem stuff is, in the end, artifacts).

Reply:
>I'm having trouble arguing that globally trusted material for the ecosystem is distinct from globally trusted material for the artifacts. (Because ecosystem stuff is, in the end, artifacts).

I do feel pretty strongly that there's a meaningful difference.

- The trust root is: what material do I need to *trust my identity provider* when it says "this cert belongs to `user@example.com`"
- The key material is: *who* is responsible for the artifacts?

Put another way: you should be able to take a *single* instance of a Sigstore verifier, and use it for completely unrelated artifact verification policies. I might not know the set of trusted keys until *after* I construct the Sigstore verification client. It's a property of the *verification policy*, not the Sigstore instance.

Am willing to die on this hill :)

Reply:
Trying to digest this conversation. If I'm understanding it correctly... 

We take `cosign` as an example. 

We execute this command `$ cosign verify --key mykey.pub mybundle.bundle`.

As part of the start `cosign` reads the TUF root and uses the content to populate a `TrustRoot`.
The `TrustRoot` would for example contain: a set of `TransparencyLogInstance`s, `CertificateAuthority`s, empty set of `public_keys` and info on the trusted CT logs.
 As part of verifying the bundle, the `mykey.pub` is considered part of the policy, and so it's passed together with the `TrustRoot` and the bundle to the `verify` method (packed into the `Input` structure, the public key is added to the `ArtifactVerificationOptions`.
This way, it's clear what is the global trust root, and what we consider valid for _this_ artifact during verification. 
Is that correctly understood @znewman01 ?

If that is correct, we are arguing for the same thing, except that I added the public key into the trust root direct. Which I can maybe agree is a tad bit awkward. But what other keys are we thinking could possibly go into the `public_keys`? Having more examples is better!

If we consider another example, the policy controller. The policy controller is a long lived process, and it's expected to prepare the 'TrustRoot` upon process start. For each artifact to verify, it needs to know about a policy, and the policy of course must reference a specific key to use. 
Following the same analogy as with `cosign`, my impression is that we are just arguing where the key should be, in the `TrustRoot` or the `ArtifactVerificationOptions`, right?

My view was that all such keys (from the polices) are added into the `TrustRoot` but are fetched/filtered during policy evaluation. Your stance is that the key material should always be part of the policy and never the trust root. I'm fine with that, and I'll give that it's probably a better separation of concerns.

Let's hope we resolve this, I don't want for you to die on this hill 😆

Reply:
> Following the same analogy as with cosign, my impression is that we are just arguing where the key should be, in the TrustRoot or the ArtifactVerificationOptions, right?

Correct :)

>  My view was that all such keys (from the polices) are added into the TrustRoot but are fetched/filtered during policy evaluation. Your stance is that the key material should always be part of the policy and never the trust root. I'm fine with that, and I'll give that it's probably a better separation of concerns.

Yeah, that's my vote. I think there are many circumstances in which we won't actually know *all* keys until later on (especially in long-lived processes such as the policy controller).

Reply:
Cool, I'll update the PR tomorrow.

Reply:
Updated, resolving this.